lib/AIEDialect.cpp

```diff
@@ -511,6 +511,10 @@ template <typename... ParentOpTypes> struct HasSomeParent {
 };
 
 LogicalResult xilinx::AIE::UseLockOp::verify() {
+  // AIE.useLock may be used in a module to set the lock's default value
+  if (llvm::isa<mlir::ModuleOp>((*this)->getParentOp()))
+    return success();
+
   return HasSomeParent<xilinx::AIE::CoreOp, xilinx::AIE::MemOp,
                        xilinx::AIE::ShimDMAOp>::verifyTrait(*this);
```

Comment:
I think all of should be implemented by tablegen using ParentOneOf:  see https://mlir.llvm.org/docs/Traits/#hasparent

Reply:
This cannot be done in tablegen as the verification here requires two different kinds of parent verification.  One is the custom `HasSomeParent` which verifies parents recursively, and the other is the `isa` test for checking the *immediate* parent.

Reply:
The immediate parent check could be done in tablegen, leaving the verify method to just do things that no good tablegen support exists for.   (And HasSomeParent might be a good thing to refactor and upstream tablegen support for anyway...)

Reply:
Sorry, I’m quite new to mlir.  Please bear with me if this question sounds naive.  Does writing a trait in tablegen express an “or” semantics to the verify function?  It sounds “and” to me.  If not, how do I write it to implement: immediate parent “or” some parent.

Reply:
Ah, I see your concern now :)  yeah, maybe easier to just leave it as is...

Reply:
The closest you could get would be Or<HasSomeParent<...>, ParentOneOf<....>>